pkg/util/certificates/csr.go

```diff
@@ -38,3 +47,96 @@ func ParseCertificateRequestObject(obj *certificates.CertificateSigningRequest)
 	}
 	return csr, nil
 }
+
+// NewCertificateRequest generates a PEM-encoded CSR using the supplied private
+// key file, subject, and SANs. If the private key file does not exist, it generates a
+// new ECDSA P256 key to use and writes it to the keyFile path.
+func NewCertificateRequest(keyFile string, subject *pkix.Name, dnsSANs []string, ipSANs []net.IP) ([]byte, error) {
+	var privateKey interface{}
+
+	if _, err := os.Stat(keyFile); os.IsNotExist(err) {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
+		if err != nil {
+			return nil, err
+		}
+
+		ecdsaKey := privateKey.(*ecdsa.PrivateKey)
+		derBytes, err := x509.MarshalECPrivateKey(ecdsaKey)
+		if err != nil {
+			return nil, err
+		}
+
+		pemBlock := &pem.Block{
+			Type:  "EC PRIVATE KEY",
+			Bytes: derBytes,
+		}
+
+		err = ioutil.WriteFile(keyFile, pem.EncodeToMemory(pemBlock), os.FileMode(0600))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	keyBytes, err := ioutil.ReadFile(keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var block *pem.Block
+	var sigType x509.SignatureAlgorithm
+
+	block, _ = pem.Decode(keyBytes)
+
+	switch block.Type {
+	case "EC PRIVATE KEY":
+		privateKey, err = x509.ParseECPrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+		ecdsaKey := privateKey.(*ecdsa.PrivateKey)
+		switch ecdsaKey.Curve.Params().BitSize {
+		case 521:
+			sigType = x509.ECDSAWithSHA512
+		case 384:
+			sigType = x509.ECDSAWithSHA384
+		default:
```

Comment:
should we make sure the bitsize is explicitly 256, and add a default case that errors?